### DISCUSSION TITLE
[V5] Proper way to implement type-safe actions

### DISCUSSION THREAD

**Discussion by alexamy** [statelyai/xstate#4556] [Q&A] [answered]
Hi! Thanks for the incredible XState 5. I'm currently in a migration process, and dropped the typegen in favour of `setup`. But there is a problem with typings of actions.

Currently, `setup.types.actions` is not provided, but as [the help](https://stately.ai/docs/actions#typescript) says, I can use `createMachine.types.actions` for this:
```typescript
const machine = setup({
  // actions: {
  //   'inner action': () => {},
  //   'outer action': () => {},
  // },
}).createMachine({
  types: {} as {
    actions:
      | { type: 'inner action', params: {} }
      | { type: 'outer action', params: {} }
  },
  id: 'test machine',
  entry: [
    { type: 'inner action' },
  ],
});
``` 

Problems:
1. Every action type demands `params`, even if it is empty
2. There is no autocompletion of action names in machine config

The solution is to provide empty implementation in `setup.actions` (commented in the code sample above) and remove typings.

Questions:
1. Will `setup.types.actions` be implemented in the future?
2. Is it a bug that `createMachine.types.actions` does not provide autocompletion of action names in the machine config?
3. Will the mandatory `params` be optional in the future?
4. How to get rid of empty action implementation in `setup.actions`, but preserve type safety in the machine config? (Current use case for this are actions which must be implemented in `provide` of the code that starts the machine).

**Comment by Andarist** [comment, marked as answer]
> Currently, setup.types.actions is not provided, but as [the help](https://stately.ai/docs/actions#typescript) says, I can use createMachine.types.actions for this:

This is misleading and we should correct this. You can only use `createMachine.types` when using the "old" `createMachine`. With `setup` you are supposed to "preconfigure" everything within the `setup` call.

> Every action type demands params, even if it is empty

This works OK:

```ts
setup({
  actions: {
    'inner action': () => {},
    'outer action': () => {},
  },
}).createMachine({
  entry: [{ type: 'inner action' }, 'outer action']
});
```

> There is no autocompletion of action names in machine config

That's likely because you are not configuring those actions in the `setup` but rather trying to do it within `createMachine`. With the sample above I get perfect autocompletes :)

> Will setup.types.actions be implemented in the future?

Why would you like to do that? The whole point of `setup` is that you don't have to do it - `actions` types are derived from the implementations that you provide there.

> Is it a bug that createMachine.types.actions does not provide autocompletion of action names in the machine config?

It would be a bug~ if that happened with bare `createMachine` but as mentioned, when you are using `setup` then you should provide all of the types and implementation there and we use *those* types from the `setup` to type the machine config.

> Will the mandatory params be optional in the future?

I think that you only see them as required here because you try to specify that type within the `setup.createMachine`. I'll how we can improve this. Perhaps we could just do this within this `createMachine` call: `types?: never`

> How to get rid of empty action implementation in setup.actions, but preserve type safety in the machine config? (Current use case for this are actions which must be implemented in provide of the code that starts the machine).

To reiterate what I mentioned above. The answer is - you can't. That's a big drive behind the setup API, that you have to provide all of the implementations upfront (at least those that u'd like to refer to by their string types in the config). Otherwise, you end up with a machine that can't truly be executed in full - something that is a common mistake. People could add an action type without providing the required implementation thus leading to problems down the line. While with `actions` you could argue that we can just treat them as `noop`... what should happen with missing guards? Should we treat them as `() => true`? Or as `() => false`? What about delays? Should we "default" them to 0? Or to `Infinity`? :p 

Most of the time not providing an implementation for the action that you have used is a problem worth fixing. With the `setup` API you are forced to do that *early*.

**Reply by alexamy** [reply]
Ok. Thanks for an answer.  I should rethink my understanding of machine creation. :)
Sorry for a lot of questions in one discussion. To summarize:
1. Add implementations in the `setup`, even if they are empty. You have to provide all of the implementations upfront.
2. Add typings in `setup.types` for everything else.
